documentation/general/dotnet-run-file.md

```diff
@@ -0,0 +1,283 @@
+See [the corresponding language proposal][pound].
+
+If these directives were limited by the language to only appear near the top of the file (similar to `#define` directives),
+the dotnet CLI could be more efficient in searching for them.
+
+It should be also possible to look for these directives from the dotnet CLI via a regex instead of parsing the whole C# file via Roslyn.
+
+We do not limit `#package` directives to appear only in entry point files.
+Indeed, it might be beneficial to let a non-entry-point file like `Util.cs` be self-contained and have all the `#package`s it needs specified in it,
+which also makes it possible to share it independently or symlink it to multiple script folders.
+This is also similar to `global using`s which users usually put into a single file but don't have to.
+
+We could consider deduplicating `#package` directives (if they have the same version)
+so separate "self-contained" utilities can reference overlapping sets of packages
+even if they end up in the same compilation.
+
+During [grow up](#grow-up), `#package` directives are removed from the `.cs` files and turned into `<PackageReference>` elements in the corresponding `.csproj` files.
+For project-based programs, `#package` directives are an error (reported by Roslyn when it's told it is in "project-based" mode).
+
+## SDK directive
+
+We could also recognize `#sdk` directive to allow web file-based programs for example.
+
+```cs
+#sdk Microsoft.NET.Sdk.Web
+```
+
+It should have similar restrictions as the `#package` directive.
+It should also be an error to specify multiple different `#sdk` directives
+but it could be allowed to specify the same SDK multiple times similarly to `#package` directives
+(again so that self-contained utility files can declare their required SDK).
+
+## Shebang
+
+Along with `#package`, the language can also ignore `#!` which could be then used for [shebang][shebang] support.
+
+```cs
+#!/usr/bin/dotnet run
+Console.WriteLine("Hello");
+```
+
+It might be beneficial to also ship `dotnet-run` binary
+(or `dotnet-run-file` that would only work with file-based programs, not project-based ones, perhaps simply named `cs`)
+because some shells do not support multiple command-line arguments in the shebang
+which is needed if one wants to use `/usr/bin/env` to find the `dotnet` executable
+(although `-S` argument can be sometimes used to enable multiple argument support):
+
+```cs
+#!/usr/bin/env dotnet run
+// ^ Might not work in all shells. "dotnet run" might be passed as a single argument to "env".
+```
+```cs
+#!/usr/bin/env dotnet-run
+// ^ Should work in all shells.
+```
+```cs
+#!/usr/bin/env -S dotnet run
+// ^ Workaround in some shells.
+```
+
+We could also consider making `dotnet file.cs` work because `dotnet file.dll` also works today
+but that would require changes to the native dotnet host.
+
+## Other commands
+
+We can consider supporting other commands like `dotnet build`, `dotnet pack`, `dotnet watch`.
```

Comment:
I think we should also _consider_ adding a `dotnet compile` command that takes a file (cannot take a project) but otherwise does the same thing as dotnet build. I only suggest this because I think the number of software engineers who know they need to "compile" code greatly exceeds the number who would think to "build" it.

Reply:
Yeah that's an idea that's come up in the past too and one I quite like. We can capture that here for further exploration down the road.